pkg/health/server/prober.go

```diff
@@ -228,45 +238,42 @@ func (p *prober) httpProbe(node string, ip string, port int) *models.Connectivit
 }
 
 func (p *prober) runHTTPProbe() {
+	nodes := make(map[string]*net.IPAddr)
 	p.RLock()
-	nodes := p.nodes
-	p.RUnlock()
-
-	for _, node := range nodes {
-		for elem := range getNodeAddresses(node) {
-			if skipAddress(elem) {
-				continue
+	for _, node := range p.nodes {
+		for elem, primary := range getNodeAddresses(node) {
+			if name, addr := resolveIP(node, elem, "icmp", primary); addr != nil {
+				nodes[name] = addr
 			}
+		}
+	}
+	p.RUnlock()
 
-			status := &models.PathStatus{}
-			ports := map[int]**models.ConnectivityStatus{
-				defaults.HTTPPathPort: &status.HTTP,
-			}
-			for port, result := range ports {
-				*result = p.httpProbe(node.Name, elem.IP, port)
-				if status.HTTP.Status != "" {
-					log.WithFields(logrus.Fields{
-						logfields.NodeName: node.Name,
-						logfields.IPAddr:   elem.IP,
-						logfields.Port:     port,
-					}).Debugf("Failed to probe: %s", status.HTTP.Status)
-				}
+	for name, ip := range nodes {
+		status := &models.PathStatus{}
+		ports := map[int]**models.ConnectivityStatus{
+			defaults.HTTPPathPort: &status.HTTP,
+		}
+		for port, result := range ports {
+			*result = p.httpProbe(name, ip.String(), port)
+			if status.HTTP.Status != "" {
+				log.WithFields(logrus.Fields{
+					logfields.NodeName: name,
+					logfields.IPAddr:   ip.String(),
+					logfields.Port:     port,
+				}).Debugf("Failed to probe: %s", status.HTTP.Status)
 			}
-
-			p.Lock()
-			p.results[ipString(elem.IP)].HTTP = status.HTTP
-			p.Unlock()
 		}
+
+		p.Lock()
+		p.results[ipString(ip.String())].HTTP = status.HTTP
```

Comment:
If `p.results[ipString(ip.String())]` returns nil, because it can return nil if `ipString(ip.String())` doesn't exist on the map, you get a panic for trying to access a nil pointer.

Reply:
The only time that this callback should be called is based on an IP that was added to the pinger via `p.AddIPAddr()`. In the same critical section, we always initialize the same IP's entry in the results map. 

https://github.com/cilium/cilium/blob/664b02baca1d6e948f78e4a671a2bb131c2b6619/pkg/health/server/prober.go#L177-L188

So I think it should be good. Let me know if you still think it's insufficient and I'll just add a check.